### PR TITLE
Insert a rate limit for cold backends

### PR DESCRIPTION
**Pull request by NiccoMlt**
# Desired 

We need to model the spectrum between down backends and up backends.
We need some way to describe a backend that is "cold", meaning it was down and it got up since less than some, configurable, amount of time.

Add a new configuration should be added to define how much time should pass for a cold backend to warm up; for simplicity I will add it globally, and not per backend.

Also, a new configuration should be added to define how many connections a cold backend is able to handle; this should be a per-backend configuration instead...

# Implementation

- Added configuration `healthmanager.warmupperiod` (default **`30`s**) → time needed to go from `COLD` to `STABLE`
- Added configuration `healthmanager.tolerant` (default **`false`**) → if true, admits connections even beyond `safeCapacity` if needed
- Added configuration backends.<id>.safeCapacity (default **`10` connections**) → number of connections allowed by the backend when `COLD`

### CODE REVIEW

File: carapace-server/src/main/java/org/carapaceproxy/server/backends/BackendHealthStatus.java [hunk 1]
```diff
@@ -19,75 +19,122 @@
...
+        /**
+         * The backend is reachable, but not since long.
+         * When in this safe, it is reasonable to assume that it is still warming-up,
+         * so it would be a sensible decision not to overload it with requests.
+         */
+        COLD,
+        /**
+         * The backend is reachable and was so since a reasonably-long time.
+         */
+        WARM
```
Comment (hamadodene) [inline carapace-server/src/main/java/org/carapaceproxy/server/backends/BackendHealthStatus.java:138]: If the backend is relatively stable for a sufficiently long time, perhaps we could use something like:
ACTIVE, HEALTHY, or STABLE.
In this case, STABLE might sound better as the focus is on stability over time.

What do you think @NiccoMlt  ?

Comment (NiccoMlt) [reply carapace-server/src/main/java/org/carapaceproxy/server/backends/BackendHealthStatus.java:138]: I agree, I was out of ideas on which name to use here ahah

Comment (NiccoMlt) [reply carapace-server/src/main/java/org/carapaceproxy/server/backends/BackendHealthStatus.java:138]: So we want to call them:

- `DOWN`
- `COLD`
- `STABLE`

Right?

Comment (hamadodene) [reply carapace-server/src/main/java/org/carapaceproxy/server/backends/BackendHealthStatus.java:138]: Yes, can go for me

**Review by hamadodene — APPROVED** [review]
LGTM... @dmercuriali  can you also take a look?

**Comment by NiccoMlt** [comment]
Thank you for the review @hamadodene.
I'm still reworking a bunch of things, like moving out the JSpecify introduction on a dedicated PR and improving the backend selection logic.
I think i will work on it until wednesday, then we can integrate whatever is up on this branch

File: carapace-server/src/main/java/org/carapaceproxy/client/EndpointConnection.java [hunk 2]
```diff
@@ -22,6 +22,7 @@
 import io.netty.handler.codec.http.HttpContent;
 import io.netty.handler.codec.http.HttpRequest;
 import io.netty.handler.codec.http.LastHttpContent;
+import org.carapaceproxy.core.EndpointKey;
```
Comment (dmercuriali) [inline carapace-server/src/main/java/org/carapaceproxy/client/EndpointConnection.java:25]: is this necessary?

File: carapace-server/src/main/java/org/carapaceproxy/client/ConnectionsManagerStats.java [hunk 3]
```diff
@@ -21,6 +21,7 @@
 
 import java.util.Map;
 import org.carapaceproxy.EndpointStats;
+import org.carapaceproxy.core.EndpointKey;
```
Comment (dmercuriali) [inline carapace-server/src/main/java/org/carapaceproxy/client/ConnectionsManagerStats.java:24]: is this necessary?

File: carapace-server/src/main/java/org/carapaceproxy/client/ConnectionsManager.java [hunk 4]
```diff
@@ -19,6 +19,7 @@
  */
 package org.carapaceproxy.client;
 
+import org.carapaceproxy.core.EndpointKey;
```
Comment (dmercuriali) [inline carapace-server/src/main/java/org/carapaceproxy/client/ConnectionsManager.java:22]: is this necessary?

**Review by dmercuriali — CHANGES_REQUESTED** [review]
Most of this PR is refactoring and restyling. 

We should split this in 2 PR, one with refactors and one with the actual implementation.

**Comment by NiccoMlt** [comment]
I do not agree that **most** of it is refactor, but yes, it was maybe more than strictly necessary.

I tried to stay neat with the commits, I can try to split up the PR

**Comment by NiccoMlt** [comment]
Merge after #516

**Review by dmercuriali — CHANGES_REQUESTED** [review]
needs a rebase

**Review by dmercuriali — APPROVED** [review]

**Pull request merged**